src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(last, value, null);
+        if (isEmpty()) {
+            first = newNode;
+        } else {
+            last.next = newNode;
+        }
+        last = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+        if (!isEmpty() && index == 0) {
+            Node<T> newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+            size++;
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> prevNode = findNodeByIndex(index).prev;
+            Node<T> nextNode = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(prevNode, value, nextNode);
+            prevNode.next = newNode;
+            nextNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T result = findNodeByIndex(index).item;
+        findNodeByIndex(index).item = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node;
+        for (node = first; node != null; node = node.next) {
+            if (node.item == object || object != null && object.equals(node.item)) {
+                unlink(node);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean checkElementIndex(int index) {
+        return index <= size && index >= 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index == size || !checkElementIndex(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> node;
+        if (index < (size >> 1)) {
+            node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
```

Comment:
ми можемо видалити else і писати код відразу після блоку if
```suggestion
        } 
```